lib/mayaUsd/ufe/UsdUndoMaterialCommands.cpp

```diff
@@ -0,0 +1,374 @@
+        shaderNodeDef, materialItem, shaderNodeDef->GetFamily().GetString());
+    _createShaderCmd->execute();
+    if (!_createShaderCmd->insertedChild()) {
+        TF_RUNTIME_ERROR(
+            "Could not create %s shader under %s",
+            _nodeId.c_str(),
+            _createMaterialCmd->newUfePath().string().c_str());
+        return;
+    }
+    //
+    // 4. Connect the Shader to the material:
+    //
+    if (!connectShaderToMaterial(
+            _createShaderCmd->insertedChild(), _createMaterialCmd->newPrim())) {
+        return;
+    }
+    //
+    // 5. Bind the material to the parent primitive:
+    //
+    _bindCmd = std::make_shared<BindMaterialUndoableCommand>(
+        parentItem->prim(), materialItem->prim().GetPath());
+    _bindCmd->execute();
+
+    // Successfully completed the task:
+    guard.markAsCompleted();
+}
+
+void UsdUndoAssignNewMaterialCommand::undo()
+{
+    // Reverse order of creation/execution:
+    if (_bindCmd) {
+        _bindCmd->undo();
+    }
+    // We do not need to undo the connection of the shader to the surface output
+    // We do not need to undo the creation of the surface shader output
+    if (_createShaderCmd) {
+        _createShaderCmd->undo();
+    }
+    if (_createMaterialCmd) {
+        _createMaterialCmd->undo();
+    }
+    if (_createScopeCmd) {
+        _renameScopeCmd->undo();
+        _createScopeCmd->undo();
+    }
+}
+
+void UsdUndoAssignNewMaterialCommand::redo()
+{
+    if (_createScopeCmd) {
+        _createScopeCmd->redo();
+        _renameScopeCmd->redo();
+    }
+    _createMaterialCmd->redo();
+    _createShaderCmd->redo();
+    connectShaderToMaterial(_createShaderCmd->insertedChild(), _createMaterialCmd->newPrim());
+    _bindCmd->redo();
+}
+
+bool UsdUndoAssignNewMaterialCommand::connectShaderToMaterial(
+    Ufe::SceneItem::Ptr shaderItem,
+    UsdPrim             materialPrim)
+{
+    auto                 shaderUsdItem = std::dynamic_pointer_cast<UsdSceneItem>(shaderItem);
+    auto                 shaderPrim = UsdShadeShader(shaderUsdItem->prim());
+    UsdShadeOutput       materialOutput;
+    PXR_NS::SdrRegistry& registry = PXR_NS::SdrRegistry::GetInstance();
+    PXR_NS::SdrShaderNodeConstPtr shaderNodeDef
+        = registry.GetShaderNodeByIdentifier(TfToken(_nodeId));
+    SdrShaderPropertyConstPtr shaderOutputDef;
+    if (shaderNodeDef->GetSourceType() == "glslfx") {
+        materialOutput = UsdShadeMaterial(materialPrim).CreateSurfaceOutput();
+        shaderOutputDef = shaderNodeDef->GetShaderOutput(TfToken("surface"));
+    } else {
+        if (shaderNodeDef->GetOutputNames().size() != 1) {
+            TF_RUNTIME_ERROR(
+                "Cannot resolve which output of shader %s should be connected to surface",
+                _nodeId.c_str());
+            return false;
+        }
+        materialOutput
+            = UsdShadeMaterial(materialPrim).CreateSurfaceOutput(shaderNodeDef->GetSourceType());
```

Comment:
Why do that if there is nothing to be connected to those outputs?

Reply:
I agree that the material node won't be connected to those outputs, but can't a user use the editor to connect shaders to displacement if you created it?

Reply:
That would be the editor's responsibility to either automatically create the right ports, or offer options to create them when dealing with displacement shaders. The editor should not dictate what this command is doing.

Reply:
Displacement and volume attributes appear in your editor because UFE will allow browsing unauthored attributes. They do not exist in the UFE stage until you explicitly set their values or connect them.

The outputs for MaterialX context are another problem, because they are not in the default schema and escape UFE detection. But this is an issue for a later time, when we add knowledge on how to connect to a material port. I suspect the attempt to connect to outputs:displacement will be redirected to outputs:mtlx:displacement directly in the MayaUSD implementation of the connect command after looking at the return value of GetSourceType() on the shader we will be trying to connect to the material.